src/main.cpp

```diff
@@ -1933,249 +1936,144 @@ void MainFrame::destroy_src(void)
-        else
-            inputChannels1 = 2;
-
-        // Grab the info for the FreeDV->speaker device as well and ensure we're using
-        // the smallest number of common channels (1 or 2).
-        deviceInfo1b = Pa_GetDeviceInfo(g_soundCard1OutDeviceNum);
-        if (deviceInfo1b == NULL) {
-            wxMessageBox(wxT("Couldn't get output device info from Port Audio for Sound Card 1"), wxT("Error"), wxOK);
-            delete m_rxInPa;
-            if(two_rx)
-				delete m_rxOutPa;
-            m_RxRunning = false;
-            Pa_Terminate();
+            
+            engine->stop();
+            engine->setOnEngineError(nullptr, nullptr);
+            
             return;
         }
-        if (deviceInfo1b->maxOutputChannels == 1)
-            outputChannels1 = 1;
-        else
-            outputChannels1 = 2;
-        
-        if(two_rx) {        
-            initPortAudioDevice(m_rxInPa, g_soundCard1InDeviceNum, paNoDevice, 1,
-                            g_soundCard1SampleRate, inputChannels1, inputChannels1);
-            initPortAudioDevice(m_rxOutPa, paNoDevice, g_soundCard1OutDeviceNum, 1,
-                            g_soundCard1SampleRate, outputChannels1, outputChannels1);
-		}
-        else
-        {                    
-            initPortAudioDevice(m_rxInPa, g_soundCard1InDeviceNum, g_soundCard1OutDeviceNum, 1,
-                            g_soundCard1SampleRate, inputChannels1, outputChannels1);
-        }
-        
-        // Init Sound Card 2 ------------------------------------------------
-
-        if (g_nSoundCards == 2) {
-
-            m_txInPa = new PortAudioWrap();
-            if(two_tx)
-                m_txOutPa = new PortAudioWrap();
-            else
-                m_txOutPa = m_txInPa;
-
-            // sanity check on sound card device numbers
-
-            //printf("m_txInPa->getDeviceCount(): %d\n", m_txInPa->getDeviceCount());
-            //printf("g_soundCard2InDeviceNum: %d\n", g_soundCard2InDeviceNum);
-            //printf("g_soundCard2OutDeviceNum: %d\n", g_soundCard2OutDeviceNum);
-
-            if ((m_txInPa->getDeviceCount() <= g_soundCard2InDeviceNum) ||
-                (m_txOutPa->getDeviceCount() <= g_soundCard2OutDeviceNum)) {
-                wxMessageBox(wxT("Sound Card 2 not present"), wxT("Error"), wxOK);
-                delete m_rxInPa;
-                if(two_rx)
-                    delete m_rxOutPa;
-                delete m_txInPa;
-                if(two_tx)
-                    delete m_txOutPa;
-                m_RxRunning = false;
-                Pa_Terminate();
-                return;
+        else if (g_nSoundCards == 1)
+        {
+            // RX-only setup.
+            // Note: we assume 2 channels, but IAudioEngine will automatically downgrade to 1 channel if needed.
+            rxInSoundDevice = engine->getAudioDevice(std::string(wxGetApp().m_soundCard1InDeviceName.ToUTF8()), IAudioEngine::IN, g_soundCard1SampleRate, 2);
```

Comment:
OK, that's a neat trick!

Reply:
It's something handled on the AudioEngine side, so there'll need to be similar logic internally for both PulseAudio and PortAudio. I'm not sure if there's a way to not have to duplicate such logic, but at least it'll be contained to AudioEngine and its classes.